File: problem_037.py
#!/usr/bin/env python

print("Project Euler - Problem 37:")
print("Find the sum of the only eleven primes that are both truncatable from left to right and right to left.\n")

limit = 1000000
primes = [0, 0] + [1] * (limit-2)
count = 11
sum = 0

for n in range(len(primes)):
    # Find primes using sieve
    if primes[n] == 1:
        primes[n] = n
        m = 2
        while m * n < limit:
            primes[m * n] = 0
            m += 1

        # Look for truncatability
        found = True
        nstr = str(n)
        for i in range(1, len(nstr)):
            if primes[int(nstr[i:])] == 0:
                found = False
                break
            if primes[int(nstr[:i])] == 0:
                found = False
                break

        # Record results
        if found == True and n > 10:
            sum += n
            count -= 1
            if count == 0:
                break

print(sum)

File: problem_037.js
#!/usr/bin/node

console.log("Project Euler - Problem 37:");
console.log("Find the sum of the only eleven primes that are both truncatable from left to right and right to left.\n");

function is_truncatable(n, primes){
    // Single digit primes don't count
    if (n < 10) return false;

    // Truncate from right
    var right = n;
    while (right > 0){
        if (primes[right] == 0)
            return false;
        right = Math.floor(right / 10);
    }
        
    // Truncate from left
    var left = n;
    while (left > 0){
        if (primes[left] == 0)
            return false;
        left = Number(left.toString().slice(1));
    }

    return true;
}

const limit = 1000000;
var prime_sieve = Array.from({length: limit + 1}, (v, i) => 1);
var primes = [];

// Use a sieve to find all primes under 1,000,000.
prime_sieve[0] = 0;
prime_sieve[1] = 0;
for (var n=2; n<limit; n++){
    if (prime_sieve[n] == 1){
        primes.push(n);
        let m = 2;
        while (m * n < limit){
            prime_sieve[m * n] = 0;
            m++;
        }
    }
}

// Go through each and see if they are truncatable 
var sum = 0;
for (var prime of primes)
    if (is_truncatable(prime, prime_sieve))
        sum += prime;

console.log(sum);
